include/nil/actor/container/merkle/tree.hpp

```diff
@@ -116,6 +115,82 @@ namespace nil {
                     }
                     return *ret_p;
                 }
+
+                template<
+                    typename T, std::size_t Arity, typename LeafIterator,
+                    std::enable_if_t<crypto3::hashes::is_poseidon<typename T::hash_type>::value, bool> = true
+                >
+                nil::crypto3::containers::detail::merkle_tree_impl<T, Arity> make_merkle_tree(LeafIterator first,
+                                                                                              LeafIterator last) {
+//                    We can compare it with one-thread version with multi-thread
+//                    nil::crypto3::containers::detail::make_merkle_tree<T, Arity, LeafIterator>(first, last);
+
+                    typedef T node_type;
+                    typedef typename node_type::hash_type hash_type;
+                    typedef typename node_type::value_type value_type;
+
+                    std::size_t number_leaves = std::distance(first, last);
+                    nil::crypto3::containers::detail::merkle_tree_impl<T, Arity> ret(number_leaves);
+                    std::size_t core_count = nil::actor::smp::count;
+                    ret.resize(ret.complete_size());
+
+                    auto ret_p = make_foreign(&ret);
+                    std::size_t parallels = std::min(core_count-1, number_leaves);
+                    std::vector<nil::actor::future<>> fut;
+                    for (auto i = 0; i < parallels; ++i) {
+                        auto begin = i * (number_leaves / parallels);
+                        auto end = (i + 1) * (number_leaves / parallels);
+                        if (i == parallels - 1) {
+                            end = number_leaves;
+                        }
+                        fut.emplace_back(nil::actor::smp::submit_to(i+1, [begin, end, first, p = ret_p.get()]() {
+
+                            auto itr = first;
+                            for (auto i = begin; i < end; ++i) {
+                                auto leaf = *itr;
+                                (*p)[i] = (value_type)nil::crypto3::containers::detail::generate_poseidon_leaf_hash<hash_type>(leaf);
+                                itr++;
+                            }
+                            return nil::actor::make_ready_future<>();
+                        }));
+                        first += number_leaves / parallels;
+                    }
```

Comment:
Looks like we are making the same job iterating over `first` and `i`
```suggestion
                    std::size_t node_per_shard = number_leaves / parallels;

                    for (std::size_t i = 0; i < parallels; ++i) {
                        auto begin = first + i * chunk_size;
                        auto end = (i == parallels - 1) ? last : begin + node_per_shard;

                        futures.emplace_back(nil::actor::smp::submit_to(i + 1, [begin, end, p = ret_p.get()]() {
                            for (auto itr = begin; itr != end; ++itr) {
                                auto leaf = *itr;
                                (*p)[itr - begin] = static_cast<value_type>(
                                    nil::crypto3::containers::detail::generate_poseidon_leaf_hash<hash_type>(leaf)
                                );
                            }
                            return nil::actor::make_ready_future<>();
                        }));
                    }
```

Reply:
I'm afraid to change this, since begin and end were always numbers. Most probably they can be iterators, but let's keep it as is for now.